### DISCUSSION TITLE
Proposal: ITask and ITask<T> interfaces

### DISCUSSION THREAD

**Discussion by Joy-less** [dotnet/csharplang#8908] [Language Ideas]
Currently, it can be very tricky to work with tasks in reflection. For example, if you have a `Task<T>` cast to an `object`, there is no way to get the result without reflection. Additionally, there is no way to handle `Task` and `ValueTask` at the same time. An `ITask` interface could solve both problems.

### Before
```cs
object? obj = GetObject();
if (obj is Task task) {
    if (task.GetType().IsGenericType) {
        object? result = task.GetType().GetProperty(nameof(Task<object>.Result))!.GetValue(task);
    }
}
else if (obj is ValueTask valueTask) {
    if (valueTask.GetType().IsGenericType) {
        object? result = valueTask.GetType().GetProperty(nameof(ValueTask<object>.Result))!.GetValue(valueTask);
    }
}
```

### After
```cs
object? obj = GetObject();
if (obj is ITask task) {
    if (task.TryGetResult(out object? result)) {
        // ...
    }
}
```

A [proposal](https://github.com/dotnet/csharplang/discussions/2498) already exists to make `Task` covariant, meaning `Task<string>` could be cast to `Task<object>`. However, this proposal would also handle `ValueTask`, and any other type of `Task` added in the future.

`ITask` would look something like this:
```cs
namespace System.Threading.Tasks;

public interface ITask {
    public bool TryGetResult(out object? result);
    public bool HasResult();
    public TaskAwaiter GetAwaiter();
}

public interface ITask<T> : ITask {
    public T GetResult();
    public new TaskAwaiter<T> GetAwaiter();
}
```
Both `Task` and `ValueTask` would inherit from `ITask`. Additionally, asynchronous functions would be able to return `ITask` instead of `Task` or `ValueTask`.

**Comment by HaloFour** [comment]
This is an API request and should be made to the [runtime](https://github.com/dotnet/runtime) repository.

**Reply by Joy-less** [reply]
Apologies. It has been reopened here: https://github.com/dotnet/runtime/discussions/110930
